samples/performance/async_compute/async_compute.cpp

```diff
@@ -0,0 +1,815 @@
+	// It's here to represent a plausible compute post workload.
+	// - Threshold pass
+	// - Blur down
+	// - Blur up
+	command_buffer.bind_pipeline_layout(*threshold_pipeline);
+	dispatch_pass(*blur_chain_views[0], get_current_forward_render_target().get_views()[0]);
+
+	command_buffer.bind_pipeline_layout(*blur_down_pipeline);
+	for (uint32_t index = 1; index < blur_chain_views.size(); index++)
+	{
+		dispatch_pass(*blur_chain_views[index], *blur_chain_views[index - 1]);
+	}
+
+	command_buffer.bind_pipeline_layout(*blur_up_pipeline);
+	for (uint32_t index = blur_chain_views.size() - 2; index >= 1; index--)
+	{
+		dispatch_pass(*blur_chain_views[index], *blur_chain_views[index + 1], index == 1);
+	}
+
+	// We're going to read the HDR texture again in the present queue.
+	// Need to release ownership back to that queue.
+	if (post_compute_queue->get_family_index() != present_graphics_queue->get_family_index())
+	{
+		vkb::ImageMemoryBarrier memory_barrier{};
+		memory_barrier.old_layout       = VK_IMAGE_LAYOUT_SHADER_READ_ONLY_OPTIMAL;
+		memory_barrier.new_layout       = VK_IMAGE_LAYOUT_SHADER_READ_ONLY_OPTIMAL;
+		memory_barrier.src_access_mask  = 0;
+		memory_barrier.dst_access_mask  = 0;
+		memory_barrier.src_stage_mask   = VK_PIPELINE_STAGE_COMPUTE_SHADER_BIT;
+		memory_barrier.dst_stage_mask   = VK_PIPELINE_STAGE_BOTTOM_OF_PIPE_BIT;
+		memory_barrier.old_queue_family = post_compute_queue->get_family_index();
+		memory_barrier.new_queue_family = present_graphics_queue->get_family_index();
+
+		command_buffer.image_memory_barrier(get_current_forward_render_target().get_views()[0], memory_barrier);
+	}
+
+	command_buffer.end();
+
+	VkPipelineStageFlags wait_stages[]     = {VK_PIPELINE_STAGE_COMPUTE_SHADER_BIT, VK_PIPELINE_STAGE_COMPUTE_SHADER_BIT};
+	VkSemaphore          wait_semaphores[] = {wait_graphics_semaphore, wait_present_semaphore};
+	VkSemaphore          signal_semaphore  = render_context->request_semaphore();
+
+	auto info                 = vkb::initializers::submit_info();
+	info.pSignalSemaphores    = &signal_semaphore;
+	info.signalSemaphoreCount = 1;
+	info.pWaitSemaphores      = wait_semaphores;
+	info.waitSemaphoreCount   = wait_present_semaphore != VK_NULL_HANDLE ? 2 : 1;
+	info.pWaitDstStageMask    = wait_stages;
+	info.commandBufferCount   = 1;
+	info.pCommandBuffers      = &command_buffer.get_handle();
+
+	if (wait_present_semaphore != VK_NULL_HANDLE)
+	{
+		render_context->release_owned_semaphore(wait_present_semaphore);
+	}
+
+	queue.submit({info}, VK_NULL_HANDLE);
+	return signal_semaphore;
+}
+
+void AsyncComputeSample::update(float delta_time)
+{
+	if (last_async_enabled != async_enabled)
+	{
+		setup_queues();
+	}
+
+	// We can potentially get more overlap if we double buffer the HDR render target.
+	// In this scenario, the next frame can run ahead a little further before it needs to block.
+	if (double_buffer_hdr_frames)
```

Comment:
I think you need an 'else' here to reset forward_render_target_index to 0 when switching out of double buffer mode.
Double-buffer mode gives a 'lumpy' framerate (faster then slower every second or so). I noticed that when switching double-buffer mode off, the lumpiness continued.  Adding the 'else' case fixes the issue for me.

Reply:
Not exactly sure why, but it's somewhat cleaner.